x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/WaitUntilReplicateForTimePassesStepTests.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ilm;
+
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.LifecycleExecutionState;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexVersion;
+import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xpack.core.ilm.step.info.EmptyInfo;
+
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+import static org.elasticsearch.cluster.metadata.LifecycleExecutionState.ILM_CUSTOM_METADATA_KEY;
+import static org.elasticsearch.xpack.core.ilm.WaitUntilReplicateForTimePassesStep.approximateTimeRemaining;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+public class WaitUntilReplicateForTimePassesStepTests extends AbstractStepTestCase<WaitUntilReplicateForTimePassesStep> {
+
+    @Override
+    protected WaitUntilReplicateForTimePassesStep createRandomInstance() {
+        Step.StepKey stepKey = randomStepKey();
+        Step.StepKey nextStepKey = randomStepKey();
+        TimeValue replicateFor = randomPositiveTimeValue();
+        return new WaitUntilReplicateForTimePassesStep(stepKey, nextStepKey, replicateFor, Instant::now);
+    }
+
+    @Override
+    protected WaitUntilReplicateForTimePassesStep mutateInstance(WaitUntilReplicateForTimePassesStep instance) {
+        Step.StepKey key = instance.getKey();
+        Step.StepKey nextKey = instance.getNextStepKey();
+        TimeValue replicateFor = randomTimeValue();
+
+        switch (between(0, 2)) {
+            case 0 -> key = new Step.StepKey(key.phase(), key.action(), key.name() + randomAlphaOfLength(5));
+            case 1 -> nextKey = new Step.StepKey(nextKey.phase(), nextKey.action(), nextKey.name() + randomAlphaOfLength(5));
+            case 2 -> replicateFor = randomValueOtherThan(replicateFor, ESTestCase::randomPositiveTimeValue);
+        }
+        return new WaitUntilReplicateForTimePassesStep(key, nextKey, replicateFor, Instant::now);
+    }
+
+    @Override
+    protected WaitUntilReplicateForTimePassesStep copyInstance(WaitUntilReplicateForTimePassesStep instance) {
+        return new WaitUntilReplicateForTimePassesStep(
+            instance.getKey(),
+            instance.getNextStepKey(),
+            instance.getReplicateFor(),
+            Instant::now
+        );
+    }
+
+    public void testEvaluateCondition() {
+        // a mutable box that we can put Instants into
+        final Instant[] returnVal = new Instant[1];
```

Comment:
Can we use `AtomicReference` for this instead? This seems like a wallhack to me, and is less readable :)

Reply:
Absolutely, fixed in d88150a8fdd.